### PR TITLE
Add ADB Commander feature

### LINKED ISSUE #491: [Feature]: Support arbitrary ADB commands
### What do you want to see in Flocon?

Flocon already leverages ADB to power certain features. I am curious to hear your thoughts on adding a dedicated "ADB" section that would allow users to execute arbitrary ADB commands and save them for future use.

I would be happy to help implement such a feature, but I wanted to gather feedback first to ensure this aligns with Flocon's vision.

Note: Following was written in parts by AI to make a convincing case. 🙈 

---

### 🚀 Motivation
Flocon is already a powerful companion for inspecting network, databases, and analytics. However, developers often encounter scenarios where they need to modify the device state to test specific edge cases (e.g., "What happens if I turn off WiFi now?" or "I need to clear the app data to test the onboarding flow").

Currently, this requires context-switching:
- Alt-Tab away from Flocon.
- Open a Terminal
-  Type adb shell ... commands manually.
- Alt-Tab back to Flocon/Emulator.

### 💡 The Proposal
Adding a dedicated ADB Section (or "Commander") to the Flocon desktop client. This interface would allow users to:
- Execute arbitrary ADB commands directly from Flocon.
- Save frequently used commands for one-click execution.

### 🛠 Proposed Features & UI

1. The Command Runner
A simple input field to type a command (e.g., input keyevent 4) with a "Run" button. It should display the stdout/stderr output in a console-like window below.
2. "Quick Actions" / Saved Library
A grid of saved buttons. Users can create a button, name it (e.g., "Reset App"), and assign a command to it.

### ⚡ Use Cases & Examples

🔬 State & Connectivity Testing
- Network Toggling: Rapidly switch network states to test offline-first apps. 
    - svc wifi enable / svc wifi disable
    - svc data enable / svc data disable
- Airplane Mode:
    - settings put global airplane_mode_on 1 (requires broadcast)

🧹 Application Management
- Fresh Install Simulation:
    - pm clear <package_name> (Clear Data/Cache)
    - pm uninstall <package_name>
- Permissions:
    - pm grant <package_name> android.permission.CAMERA
    - pm revoke <package_name> android.permission.RECORD_AUDIO

⌨️ TV & UI Interaction (Huge for Android TV/FireTV)
Typing on a TV remote is painful.

- Text Injection:
    - input text "super_long_test_password_123"
- Key Events:
    - input keyevent 3 (Home)

📐 UI Debugging
- Accessibility:
    - settings put system font_scale 1.3
- Show Taps:
    - settings put system show_touches 1

### PR DESCRIPTION
**Pull request by aldefy**
## Summary
- Adds a new **ADB Commander** screen to the desktop app (closes #491)
- Three-tab UI: **Runner** (console output), **Library** (saved commands, automation flows, 47 built-in quick commands), **History** (timestamped command log)
- Supports **Automation Flows** - multi-step ADB command sequences with configurable delays between steps
- Works with or without a Flocon-paired device (falls back to targeting all connected ADB devices)
- Full clean architecture: domain models, Room entities, use cases, repository, ViewModel, Compose UI

## Details
- 12 use cases covering command execution, save/delete/update for commands and flows, history management, and flow execution with cancellation support
- 4 new Room entities (saved commands, history, flows, flow steps) - no FK to DeviceEntity so it works standalone
- 47 built-in quick commands across 8 categories (Device Info, App Management, Input, Connectivity, Debug, Settings, Performance)
- Flow editor dialog for creating/editing multi-step automation flows
- DB version bumped to 80

## Test plan
- [ ] Connect an Android device/emulator
- [ ] Navigate to ADB Commander in left sidebar
- [ ] Runner tab: type `shell echo hello` → Run → verify output in console
- [ ] Save command → verify it appears in Library tab
- [ ] Library tab: Run a saved command → verify execution
- [ ] Try built-in quick commands (e.g. Device Model, Battery Status)
- [ ] Create an automation flow with 2+ steps and delays
- [ ] Run flow → verify step-by-step progress in Runner tab
- [ ] Cancel flow mid-execution → verify remaining steps marked as Skipped
- [ ] History tab: verify all commands logged with timestamps
- [ ] Test without a Flocon-paired device (should still work via ADB directly)

### CODE REVIEW

**Comment by aldefy** [comment]
@florent37 Would you mind reviewing this PR? It implements #491 (ADB Commander feature).

File: .github/screenshots/history.png
Comment (rteyssandier) [inline .github/screenshots/history.png:1]: remove screenshots

File: FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/AdbCommanderScreen.kt [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import org.koin.compose.viewmodel.koinViewModel
+
+@Composable
+fun AdbCommanderScreen(modifier: Modifier = Modifier) {
+    val viewModel: AdbCommanderViewModel = koinViewModel()
+    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
+
+    AdbCommanderScreen(
+        uiState = uiState,
+        onTabSelected = viewModel::onTabSelected,
```
Comment (rteyssandier) [inline FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/AdbCommanderScreen.kt]: please use MVI instead

File: FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/RunnerContent.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            style = FloconTheme.typography.titleSmall,
+            color = FloconTheme.colorPalette.onPrimary,
+        )
+        execution.steps.forEach { step ->
+            Row(
+                modifier = Modifier.fillMaxWidth().padding(start = 8.dp),
+                horizontalArrangement = Arrangement.spacedBy(8.dp),
+                verticalAlignment = Alignment.CenterVertically,
+            ) {
+                val statusIcon = when (step.status) {
```
Comment (rteyssandier) [inline FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/RunnerContent.kt]: can you use material icons for this ? Make an extension too

**Comment by rteyssandier** [comment]
@florent37 suggested to make a menu on the left, like for Database. Would be easier to switch between commands

File: FloconDesktop/data/local/src/commonMain/kotlin/io/github/openflocon/data/local/adbcommander/datasource/LocalAdbCommanderDataSourceRoom.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    override fun observeFlows(deviceId: String): Flow<List<AdbFlowDomainModel>> =
+        dao.observeFlows(deviceId)
+            .map { entities ->
+                entities.map { flowEntity ->
+                    val steps = dao.getFlowSteps(flowEntity.id)
+                    flowEntity.toDomainModel(steps)
+                }
+            }
+            .distinctUntilChanged()
```
[the remark this answers is not in the document]
Comment (aldefy) [reply FloconDesktop/data/local/src/commonMain/kotlin/io/github/openflocon/data/local/adbcommander/datasource/LocalAdbCommanderDataSourceRoom.kt:71]: Fixed in 587e4f9 — using @Relation with AdbFlowWithSteps POJO to fetch flows with steps in a single query.

File: FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/mapper/Mapper.kt [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+fun AdbFlowExecutionState.toUiModel() = FlowExecutionUiModel(
+    flowName = flowName,
+    steps = steps.map { stepState ->
+        FlowExecutionStepUiModel(
+            label = stepState.step.label ?: stepState.step.command,
+            command = stepState.step.command,
+            status = stepState.status.name,
+            output = stepState.output,
+            isActive = stepState.status == AdbFlowExecutionState.StepStatus.Running
+                    || stepState.status == AdbFlowExecutionState.StepStatus.WaitingDelay,
```
[the remark this answers is not in the document]
Comment (aldefy) [reply FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/mapper/Mapper.kt]: Fixed in 6366822 — using in setOf() for status checks.

File: FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/mapper/Mapper.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        )
+    },
+    status = status.name,
+    isRunning = status == AdbFlowExecutionState.FlowStatus.Running,
+)
+
+private fun formatTimestamp(epochMs: Long): String {
+    val sdf = SimpleDateFormat("HH:mm:ss", Locale.getDefault())
+    return sdf.format(Date(epochMs))
+}
```
[the remark this answers is not in the document]
Comment (aldefy) [reply FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/mapper/Mapper.kt:60]: Fixed in 6366822 — hoisted to a top-level val.

File: FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/AdbCommanderViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+    private val observeFlowsUseCase: ObserveFlowsUseCase,
+    private val saveFlowUseCase: SaveFlowUseCase,
+    private val deleteFlowUseCase: DeleteFlowUseCase,
+    private val updateFlowUseCase: UpdateFlowUseCase,
+    private val executeFlowUseCase: ExecuteFlowUseCase,
+) : ViewModel() {
+
+    private val localState = MutableStateFlow(AdbCommanderUiState())
+    private var flowExecutionJob: Job? = null
+    private var domainFlows: List<AdbFlowDomainModel> = emptyList()
```
[the remark this answers is not in the document]
Comment (aldefy) [reply FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/AdbCommanderViewModel.kt]: Fixed in 865e7ef — domainFlows is now a proper StateFlow, accessed via .value.

File: FloconDesktop/data/local/src/commonMain/kotlin/io/github/openflocon/data/local/adbcommander/datasource/LocalAdbCommanderDataSourceRoom.kt [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                name = flow.name,
+                description = flow.description,
+                createdAt = System.currentTimeMillis(),
+            )
+        )
+        dao.replaceFlowSteps(
+            flowId = flow.id,
+            steps = flow.steps.map { it.toEntity(flow.id) },
+        )
+    }
```
[the remark this answers is not in the document]
Comment (aldefy) [reply FloconDesktop/data/local/src/commonMain/kotlin/io/github/openflocon/data/local/adbcommander/datasource/LocalAdbCommanderDataSourceRoom.kt:113]: Fixed in d245983 — fetching existing entity before update to preserve original createdAt.

File: .github/screenshots/history.png
Comment (aldefy) [reply .github/screenshots/history.png:1]: Fixed — screenshots removed from the repository.

File: FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/AdbCommanderScreen.kt (hunk 1 above)
Comment (aldefy) [reply FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/AdbCommanderScreen.kt]: Fixed — refactored to MVI pattern with a sealed `AdbCommanderAction` interface. All views now take a single `onAction: (AdbCommanderAction) -> Unit` callback, and the ViewModel dispatches via `when(action)`.

File: FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/RunnerContent.kt (hunk 2 above)
Comment (aldefy) [reply FloconDesktop/composeApp/src/commonMain/kotlin/io/github/openflocon/flocondesktop/features/adbcommander/view/RunnerContent.kt]: Fixed — replaced text-based indicators with Material icons (`Icons.Outlined.Check`, `Close`, `PlayArrow`, `HourglassEmpty`, `RemoveCircle`, `Schedule`) and added `String.toIcon()` / `String.toColor()` extension functions in RunnerContent.

**Comment by aldefy** [comment]
Great suggestion, will tackle the sidebar layout (like Database) as a follow-up. The current PR uses tabs, but a left panel with collapsible sections (Saved Commands, Flows, Quick Commands, History) alongside an always-visible console would be a better UX. Will open a separate issue/PR for this or would you prefer to have it planned on this PR?

**Comment by aldefy** [comment]
Just checking in if this PR needs any more changes , happy to fix/add changes

**Comment by florent37** [comment]
I like the feature honnestly, and it's easy to use, I'm just a bit concerned about the UX
I mean : 

- you don't see easily the textfield on top, maybe you should try to do something like Database 

Screenshot 2026-02-27 at 22 20 55
Screenshot 2026-02-27 at 22 22 04

- having pre configured commands is a must have, I'm agree with you, just every "line" should looks like a button IMO, and just one click to execute (instead of going on right to click on arrow to execute)

Screenshot 2026-02-27 at 22 24 17

---> again the UX of database, with a left pannel is better in term of lookalike

and it goes with my 3rd and last point : 
- eveytime you perform a "library" command, it redirects you to the "runner"

I think to fix all of these UX concerns, having the database lookalike with : 
```
                [ edittext/command ]
[ library] [ runner ]
``` 

is a good solution
happy to discuss this with you :) 

btw thanks a lot for this merge request <3  

Screenshot 2026-02-27 at 22 26 49

**Comment by aldefy** [comment]
> I like the feature honnestly, and it's easy to use, I'm just a bit concerned about the UX I mean :
> 
> * you don't see easily the textfield on top, maybe you should try to do something like Database
> 
> Screenshot 2026-02-27 at 22 20 55 Screenshot 2026-02-27 at 22 22 04
> * having pre configured commands is a must have, I'm agree with you, just every "line" should looks like a button IMO, and just one click to execute (instead of going on right to click on arrow to execute)
> 
> Screenshot 2026-02-27 at 22 24 17
> ---> again the UX of database, with a left pannel is better in term of lookalike
> 
> and it goes with my 3rd and last point :
> 
> * eveytime you perform a "library" command, it redirects you to the "runner"
> 
> I think to fix all of these UX concerns, having the database lookalike with :
> 
> ```
>                 [ edittext/command ]
> [ library] [ runner ]
> ```
> 
> is a good solution happy to discuss this with you :)
> 
> btw thanks a lot for this merge request <3
> 
> Screenshot 2026-02-27 at 22 26 49

Thanks , will push a fix for the changes today, was on a vaction  missed the email updates

**Comment by aldefy** [comment]
New commit now adds it adb commander as a dashboard 
image

**Comment by florent37** [comment]
thanks :) I'm testing and reviewing when I have time

**Comment by florent37** [comment]
better now :) you can rebase & merge when you want, I approved !

**Comment by aldefy** [comment]
Rebased , but guessing need that second review approval to merge
